mars/optimization/logical/core.py

```diff
@@ -106,17 +102,11 @@ def __init__(
         self._graph = graph
         self._records = records
         self._optimizer_cls = optimizer_cls
-
-    def __new__(
-        cls,
-        graph: EntityGraph,
-        records: OptimizationRecords,
-        optimizer_cls: Type["Optimizer"],
-    ):
-        if (cls, graph, records) in cls._instances:
-            return cls._instances[cls, graph, records]
-        inst = cls._instances[cls, graph, records] = object.__new__(cls)
-        return inst
+        self._cached_rule = functools.lru_cache(maxsize=None)(
+            lambda _rule_type: _rule_type(
+                self._graph, self._records, self._optimizer_cls
+            )
+        )
```

Comment:
It may cause memory keep growing if set `maxsize=None`, maybe we should set a fixed value.

Reply:
This is a local cache, it will be deleted with rule object.